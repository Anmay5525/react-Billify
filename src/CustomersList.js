import Table from "./Table.js";
import Loader from "./Loader.js";

const { useEffect, useState } = React;

const CustomersList = ({ handleSubRouteChange }) => {
  const [customers, setCustomers] = useState(null);

  const getCustomers = () => {
    fetch("https://rzp-training.herokuapp.com/team1/customers")
      .then((res) => res.json())
      .then((r) => setCustomers(r))
      .catch((er) => console.log(er));
  };
  useEffect(() => {
    getCustomers();
    return () => console.log("unmounted");
  }, []);
  const fields = ["name", "email", "contact", "created_at"];
  if (customers) {
    const data = customers.items.map((item) => {
      return {
        name: item.name,
        email: item.email,
        contact: item.contact,
        created_at: new Date(item.created_at * 1000).toDateString(),
      };
    });
    return (
      <div className="content">
        <div className="customers-title-container">
          <div className="customers-title">Customers</div>
          <button className="customers-new-btn" onClick={handleSubRouteChange}>
            + New Customer
          </button>
        </div>
        <Table fields={fields} data={data} />
      </div>
    );
  } else {
    return <Loader />;
  }
};
export default CustomersList;
