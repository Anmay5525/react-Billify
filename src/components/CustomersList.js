import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import Table from "./Table";
import Loader from "./Loader";
import Filter from "./Filter";

const CustomersList = ({ handleSubRouteChange }) => {
  const [customers, setCustomers] = useState(null);
  const [filter, setFilter] = useState(false);
  const [filterString, setFilterString] = useState("");

  const controller = new AbortController();
  const { signal } = controller;

  const handleFilterChange = (event) => {
    setFilterString(event.target.value);
  };

  const getCustomers = () => {
    fetch("https://rzp-training.herokuapp.com/team1/customers", { signal })
      .then((res) => res.json())
      .then((r) => setCustomers(r))
      // eslint-disable-next-line no-console
      .catch((er) => console.log(er));
  };

  useEffect(() => {
    getCustomers();
    return () => {
      controller.abort();
    };
  }, []);

  const fields = ["name", "email", "contact", "created_at"];

  if (customers) {
    const data = customers.items.map((item) => {
      return {
        name: item.name,
        email: item.email,
        contact: item.contact,
        created_at: new Date(item.created_at * 1000).toDateString(),
      };
    });

    const filteredData = data.filter((element) => {
      return element.name.toLowerCase().includes(filterString.toLowerCase());
    });

    // console.log(typeof filteredData, typeof fields);

    return (
      <div className="content">
        <div className="customers-title-container">
          <div className="customers-title">Customers</div>

          <button
            type="button"
            className="customers-new-btn"
            onClick={() => handleSubRouteChange("new")}
          >
            + New Customer
          </button>
        </div>
        <div>
          <div className="filter-container">
            {filter ? <Filter handleChange={handleFilterChange} /> : ""}
            {filter ? (
              ""
            ) : (
              <button
                type="button"
                className="simple-btn"
                onClick={() => setFilter(true)}
              >
                Filter
              </button>
            )}

            {filter ? (
              <button
                type="button"
                className="simple-btn"
                onClick={() => {
                  setFilter(false);
                  setFilterString("");
                }}
              >
                X
              </button>
            ) : (
              ""
            )}
          </div>
          <Table fields={fields} data={filteredData} />
        </div>
      </div>
    );
  }
  return <Loader />;
};

CustomersList.propTypes = {
  handleSubRouteChange: PropTypes.func.isRequired,
};

export default CustomersList;
