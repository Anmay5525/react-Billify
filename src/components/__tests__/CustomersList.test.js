import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import CustomersList from '../CustomersList';
import { BrowserRouter as Router } from "react-router-dom";

const customersListData = {
  "entity": "collection",
  "count": 2,
  "items": [
    {
      "id": "cust_Fg0tl7Y2qra6SJ",
      "entity": "customer",
      "name": "David Silva",
      "email": "s@s.com",
      "contact": "2938475867",
      "gstin": null,
      "notes": [],
      "created_at": 1600764149
    },
    {
      "id": "cust_FfeM7EaKLd8SOh",
      "entity": "customer",
      "name": "John Stones",
      "email": "s@s.com",
      "contact": "2837495867",
      "gstin": null,
      "notes": [],
      "created_at": 1600684763
    }
  ]
};

describe('Testing CustomersList component if it', () => {

  let container;

  beforeEach(async () => {
    fetch.resetMocks();
    fetch.mockResponseOnce(JSON.stringify(customersListData));
    container = mount(<Router><CustomersList /></Router>);
    await act(async () => { });
    container.update();
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<CustomersList></CustomersList>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('makes fetch calls', async () => {
    expect(container.html()).toMatchSnapshot();
    expect(fetch).toHaveBeenCalledTimes(1);
  });

  it('handles fetch errors', async () => {
    fetch.mockRejectOnce(['Error', {status: 200}]);
    const container = mount(<Router><CustomersList /></Router>);
    await act(async () => { });
    container.update();
    expect(container.html()).toMatchSnapshot();
  });

  it('displays correct button type', async () => {
    expect(container.find('.customers-new-btn').prop('type')).toEqual('button');
    expect(container.find('.simple-btn').prop('type')).toEqual('button');
  });

  it('displays correct number of table components', async () => {
    expect(container.find('table')).toHaveLength(1);
    expect(container.find('tr')).toHaveLength(3);
  });

  it('displays correct table columns', async () => {
    const tr0 = container.find('tr').at(0);
    expect(tr0.childAt(0).contains('NAME')).toEqual(true);
    expect(tr0.childAt(1).contains('EMAIL')).toEqual(true);
    expect(tr0.childAt(2).contains('CONTACT')).toEqual(true);
    expect(tr0.childAt(3).contains('CREATED AT')).toEqual(true);
  });

  it('displays correct table data', async () => {
    const tr1 = container.find('tr').at(1);
    const tr2 = container.find('tr').at(2);
    expect(tr1.childAt(0).contains('David Silva')).toEqual(true);
    expect(tr1.childAt(1).contains('s@s.com')).toEqual(true);
    expect(tr1.childAt(2).contains('2938475867')).toEqual(true);
    expect(tr1.childAt(3).contains('Tue Sep 22 2020')).toEqual(true);
    expect(tr2.childAt(0).contains('John Stones')).toEqual(true);
    expect(tr2.childAt(1).contains('s@s.com')).toEqual(true);
    expect(tr2.childAt(2).contains('2837495867')).toEqual(true);
    expect(tr2.childAt(3).contains('Mon Sep 21 2020')).toEqual(true);
  });

  it('contains button which turns filter on', async () => {
    let simpleButton = container.find('.simple-btn');
    simpleButton.simulate('click');
    const filterInput = container.find('.filter-input');
    expect(filterInput).toHaveLength(1);
    expect(filterInput.is('input')).toEqual(true);
    simpleButton = container.find('.simple-btn');
    expect(simpleButton).toHaveLength(1);
    expect(simpleButton.is('button')).toEqual(true);
    expect(simpleButton.type()).toEqual('button');
  });
  
  it('handles filter change', async () => {
    container.find('.simple-btn').simulate('click');
    container.find('.filter-input').simulate('change', {target: {value: 'Da'}});
    expect(container.find('table').text().includes('John Stones')).toEqual(false);
  });

  it('contains button which turns filter off', async () => {
    container.find('.simple-btn').simulate('click');
    container.find('.simple-btn').simulate('click');
    expect(container.find('.filter-container').contains('<button type="button" class="simple-btn">Filter</button>'));
  });

});
