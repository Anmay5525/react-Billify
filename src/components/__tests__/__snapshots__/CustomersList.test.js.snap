// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing CustomersList component if it handles fetch errors 1`] = `"<div class=\\"loader-container\\"><svg viewBox=\\"0 0 100 100\\" xmlns=\\"http://www.w3.org/2000/svg\\"><circle cx=\\"50\\" cy=\\"50\\" r=\\"45\\"></circle></svg></div>"`;

exports[`Testing CustomersList component if it makes fetch calls 1`] = `"<div class=\\"content\\"><div class=\\"customers-title-container\\"><div class=\\"customers-title\\">Customers</div><a href=\\"/Customers/New\\"><button type=\\"button\\" class=\\"customers-new-btn\\">+ New Customer</button></a></div><div><div class=\\"filter-container\\"><button type=\\"button\\" class=\\"simple-btn\\">Filter</button></div><div><table class=\\"table\\"><thead><tr><th>NAME</th><th>EMAIL</th><th>CONTACT</th><th>CREATED AT</th></tr></thead><tbody><tr class=\\"tr\\"><td>David Silva</td><td>s@s.com</td><td>2938475867</td><td>Tue Sep 22 2020</td></tr><tr class=\\"tr\\"><td>John Stones</td><td>s@s.com</td><td>2837495867</td><td>Mon Sep 21 2020</td></tr></tbody></table></div></div></div>"`;
